### DISCUSSION TITLE
dependency on X-display?

### DISCUSSION THREAD

**Discussion by zbenmo** [bokeh/bokeh#13290] [General]
Is there a dependency on X-display? Must there be an X-display available when bokeh is used?

I hope the answer is no, but I suspect that somehow the dependency creeped in.

I am using 'bulk' (Vincent Warmerdam), on a Windows machine, WS2 (Ubuntu).
Without X server on my Windows machine, I cannot proceed.

I had similar situation with matplotlib. The solution there was with mpl.use('Agg') or so.

**Comment by bryevdv** [comment]
Bokeh has no dependency on X, full stop To be clear: Bokeh generates HTML and JavaScript as output. Many people use Bokeh on headless systems to generate such output, e.g. in web servers. Of course, if you want to *view* that output, you will need to open it in a browser, which will require *some* kind of display. But I use Bokeh almost exclusively on OSX (no X there) and other users use it daily on Windows (same, no X-server). This has been the case since the project started, more than ten years ago. 

Bokeh can also generate PNG output via Selenium and a headless browser, many people also use this on headless systems as well. But you will need a compatible webdriver and browser installed, even on a headless system. 

@zbenmo Why are you asking?  Is this just a general question or unprompted speculation? Did you encounter a specific situation that prompted this question? If so, you should always provide details and context, links, information, reproducible steps, *actual complete error messages*, etc to make the situation and question understandable. 

> I am using 'bulk' (Vincent Warmerdam),

I have ***no idea*** what this is, or if it is even related to Bokeh in *any way*.  Would you care to explain what it is and provide some links to the project?

**Comment by zbenmo** [comment]
This 'bulk' utility is making use of bokeh. 'bokeh' was the main suspect. Cannot think of other "graphics" related "image generation" or so (or a dependency on a library that does, like matplotlib etc.). But okay. You don't believe it is related. Will keep searching for the issue. Thanks. Will continue to search.

**Reply by bryevdv** [reply]
It's literally impossible to speculate without (much) more information. For instance, Bokeh has a convenience function `show` that is specifically intended for interactive use that both saves the HTML/JS output to a file, *and raises a browser window to display*. Maybe "bulk" is using `show` when they could just call `save`? I could try to quickly check, but you haven't linked to a repo for the other package. I'd love to be able to help, but there here is no way to say anything at all without:

* detailed usage instructions to reproduce
* descriptions of expected behavior vs actual
* full and complete error messages
* a listing of relevant software complete with exact versions

As is it I don't even know what your issue looks like: bad output? no output? silent crash? crash with error messages? software hanging? A combination of these? Something else entirely?

You need to help others help you. Please take a look at this artice [How to create a Minimal, Reproducible Example](https://stackoverflow.com/help/minimal-reproducible-example) That is the *minimum* bar that should be met asking for help with *any* OSS project.

**Comment by zbenmo** [comment]
Okay, not related to Bokeh at least I think. 'bulk' is trying to open a browser. This is the issue. Thanks for the suggestions and apologies. BR, Oren.
